### PR TITLE
Mtracy/more js refactors

### PR DESCRIPTION
**Pull request by mrtracy**
A few commits:
1. Factor all requests to use a common Http package.
2. Finally rename models/timeseries.ts -> models/metrics.ts
3. Create the "Axis" type, and change all charts to require an Axis. This improves the efficiency of queries (more batching) while at the same time greatly improving our flexibility for labeling charts.

Three separate commits.

### CODE REVIEW

File: resource/ui/ts/models/metrics.ts [hunk 1]
```diff
@@ -146,25 +269,37 @@ module Models {
...
-                for (var i = 0; i < this._selectors.length; i++) {
-                    req.queries.push(this._selectors[i].request())
-                }
+                
+                // Uniquify requests from our selector set.
+                var requestSet = new QueryInfoSet<Proto.QueryRequest>();
+                this.selectors().forEach((s:select.Selector) => {
+                    requestSet.add(s.request());
+                });
+                requestSet.forEach((qr:Proto.QueryRequest) => {
```
Comment (BramGruneir) [inline resource/ui/ts/models/metrics.ts]: Is there a need for this second loop?

File: resource/ui/ts/models/metrics.ts [hunk 2]
```diff
@@ -146,25 +269,37 @@ module Models {
...
+                    req.queries.push(qr);
+                });
                 return Query.dispatch_query(req)
             }
 
-            private static dispatch_query(q:Proto.QueryRequestSet):promise<Proto.QueryResultSet> {
-                var url = "/ts/query";
-                return m.request({url:url, method:"POST", extract:nonJsonErrors, data:q})
+            private static dispatch_query(q:Proto.QueryRequestSet):promise<QueryResultSet> {
+                var result = new QueryInfoSet<Proto.QueryResult>();
```
Comment (BramGruneir) [inline resource/ui/ts/models/metrics.ts]: you can move this declaration into the .then

File: resource/ui/ts/models/metrics.ts [hunk 3]
```diff
@@ -186,15 +322,23 @@ module Models {
...
          */
         export function NewQuery(...selectors:select.Selector[]):Query {
-            return new Query(selectors);
+            return new Query().selectors(selectors);
         }
 
         /**
-         * nonJsonErrors ensures that error messages returned from the server
-         * are parseable as JSON strings.
+         * Executor is a convenient class for persisting the results of a query
```
Comment (BramGruneir) [inline resource/ui/ts/models/metrics.ts]: Executor is a convenience class for persisting the results of a query.

File: resource/ui/ts/pages/graph.ts [hunk 4]
```diff
@@ -22,23 +21,40 @@ module AdminViews {
...
+              private errorRate = metrics.select.AvgRate("cr.node.calls.error.1")
+                .title("Error call rate");
+
+              // Define query.
+              private query = metrics.NewQuery(
+                      this.successCount, 
+                      this.errorCount, 
+                      this.successRate, 
+                      this.errorRate
 					)
```
Comment (BramGruneir) [inline resource/ui/ts/pages/graph.ts]: some of this spacing is horrible.

File: resource/ui/ts/pages/stores.ts [hunk 5]
```diff
@@ -79,48 +73,71 @@ module AdminViews {
...
                 }
 
                 public constructor(storeId:string) {
                     this._storeId = storeId;
+                    this._query = metrics.NewQuery();
                     this._addChart(
-                        metrics.NewQuery(
-                            metrics.select.Avg(_storeMetric(storeId, "keycount")))
-                        .title("Key Count"));
+                        metrics.NewAxis(
```
Comment (BramGruneir) [inline resource/ui/ts/pages/stores.ts]: This is a lot of repeated code, can you extract to a common function?

**Comment by BramGruneir** [comment]
Overall LGTM, just some minor issues.

File: resource/ui/ts/models/metrics.ts (hunk 1 above)
Comment (mrtracy) [inline resource/ui/ts/models/metrics.ts]: yes, the first loop de-duplicates by adding requests to a set.

Comment (mrtracy) [inline resource/ui/ts/models/metrics.ts]: I'll change the comment.

File: resource/ui/ts/pages/stores.ts (hunk 5 above)
Comment (mrtracy) [inline resource/ui/ts/pages/stores.ts]: I expect these pages to change considerably as work continues, I'm not sure where the list of charts to display will end up.  I don't want to spend too much time removing boilerplate.

File: resource/ui/ts/pages/graph.ts (hunk 4 above)
Comment (mrtracy) [inline resource/ui/ts/pages/graph.ts]: Do you just mean the tabs, because this does need to have tabs -> spaces.

Comment (BramGruneir) [inline resource/ui/ts/pages/graph.ts]: We need to have some consistency for spacing.  The close bracket at the end all the way to the right seems very strange to me.

File: resource/ui/ts/models/metrics.ts (hunk 2 above)
Comment (mrtracy) [inline resource/ui/ts/models/metrics.ts]: Done.

File: resource/ui/ts/models/metrics.ts (hunk 3 above)
Comment (mrtracy) [inline resource/ui/ts/models/metrics.ts]: Done.

File: resource/ui/ts/pages/graph.ts (hunk 4 above)
Comment (mrtracy) [inline resource/ui/ts/pages/graph.ts]: That's just an effect of the tabs - I did not have the "expandtab" option on (which converts tabs to spaces), so this looks different in my editor.

I retabbed the file, it should look better now.

**Comment by mrtracy** [comment]
Addressed feedback.

File: resource/ui/ts/models/metrics.ts (hunk 1 above)
Comment (BramGruneir) [inline resource/ui/ts/models/metrics.ts]: cool

**Comment by BramGruneir** [comment]
LGTM

**Pull request merged**